src/main/java/health/tabia/challenge/MetricList.java

```diff
@@ -0,0 +1,115 @@
+package health.tabia.challenge;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class MetricList implements MetricStore {
+
+    // It was used list to handle indexes and iterations needed to remove an
+    // object
+    private List<Metric> store;
+    private int size;
+
+    public MetricList(List<Metric> newStore) {
+        this.store = newStore;
+        this.size = newStore.size();
+    }
+
+    public int getSize() {
+        return this.size;
+    }
+
+    public MetricIterator iterator() {
+        MetricIterator metricIterator = new MetricIterator() {
+
+            private int currentIndex = 0;
+
+            @Override
+            public void close() throws Exception {
+                System.out.println("Closing metric iterator...");
+            }
+
+            @Override
+            public boolean moveNext() {
+                currentIndex += 1;
+                return (currentIndex < size && store.get(currentIndex) != null);
+            }
+
+            @Override
+            public void remove() {
+                store.remove(currentIndex);
+                size--;
+            }
+
+            @Override
+            public Metric current() {
+                return store.get(currentIndex);
+            }
+
+        };
+
+        return metricIterator;
+    }
+
+    @Override
+    synchronized public void insert(Metric metric) {
+        if (this.store == null) {
+            this.store = new ArrayList<Metric>();
+        }
+
+        this.store.add(metric);
+        this.size++;
+    }
+
+    @Override
+    synchronized public void removeAll(String name) {
+        List<Metric> cleanedList = new ArrayList<>();
+        if (this.store != null) {
+            for (Metric metric : this.store) {
+                if (!metric.getName().equalsIgnoreCase(name)) {
+                    cleanedList.add(metric);
+                }
+            }
+            this.store = cleanedList;
+            this.size = cleanedList.size();
+        }
+    }
+
+    @Override
+    synchronized public MetricIterator query(String name, long from, long to) {
```

Comment:
This operation (and also `removeAll`) can be way faster. How could you do that?

Reply:
I think this is possible using filter. I gonna to do that.

Reply:
I have changed.

Reply:
The code is correct, but the performance still can be improved *a lot* :)

Reply:
Can you tell what is the complexity of the `query` method's current implementation?